src/payment/strategies/apple-pay/apple-pay-payment-strategy.ts

```diff
@@ -0,0 +1,189 @@
+                'visa',
+                'masterCard',
+                'amex',
+                'discover',
+            ],
+            lineItems,
+            total: {
+                label: storeName,
+                amount: `${checkout.grandTotal.toFixed(cart.currency.decimalPlaces)}`,
+                type: 'final',
+            },
+        };
+    }
+​
+    private _handleApplePayEvents(applePaySession: ApplePaySession, paymentMethod: PaymentMethod, promise: ApplePayPromise) {
+        applePaySession.onvalidatemerchant = async event => {
+            try {
+                const { body: merchantSession } = await this._onValidateMerchant(paymentMethod, event);
+                applePaySession.completeMerchantValidation(merchantSession);
+            } catch (err) {
+                throw new Error('Merchant validation failed');
+            }
+        };
+​
+        applePaySession.oncancel = async () =>
+            promise.reject(new PaymentMethodCancelledError('Continue with applepay'));
+​
+        applePaySession.onpaymentauthorized = (event: ApplePayJS.ApplePayPaymentAuthorizedEvent) =>
+            this._onPaymentAuthorized(event, applePaySession, paymentMethod, promise);
+    }
+​
+    private async _onValidateMerchant(paymentData: PaymentMethod, event: ApplePayJS.ApplePayValidateMerchantEvent) {
+        const body = [
+            `validationUrl=${event.validationURL}`,
+            `merchantIdentifier=${paymentData.initializationData.merchantId}`,
+            `displayName=${paymentData.initializationData.storeName}`,
+            `domainName=${window.location.hostname}`,
+        ].join('&');
+​
+        return this._requestSender.post(validationEndpoint(paymentData.initializationData.paymentsUrl), {
+            credentials: false,
+            headers: {
+                Accept: 'application/json',
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'X-XSRF-TOKEN': null,
+            },
+            body,
+        });
+    }
+​
+    private async _onPaymentAuthorized(
+        event: ApplePayJS.ApplePayPaymentAuthorizedEvent,
+        applePaySession: ApplePaySession,
+        paymentMethod: PaymentMethod,
+        promise: ApplePayPromise
+    ) {
+        const { token } = event.payment;
+        const payment: Payment = {
+            methodId: paymentMethod.id,
+            paymentData: {
+                formattedPayload: {
+                    apple_pay_token: {
+                        payment_data: token.paymentData,
+                        payment_method: token.paymentMethod,
+                        transaction_id: token.transactionIdentifier,
+                    },
+                },
+            },
+        };
+
+        try {
+            await this._store.dispatch(this._paymentActionCreator.submitPayment(payment));
+            applePaySession.completePayment(ApplePaySession.STATUS_SUCCESS);
+
+            return promise.resolve(this._store.getState());
```

Comment:
This is required here, since we return a promise as part of the `#execute` method, so when the payment fails/succeeds then we need to resolve/reject that promise.

Reply:
Ah yes, you're right, I misread `promise.resolve` vs `Promise.resolve`. 👍